File: packages/api-models-py/skyportal_py_models/observations.py
"""Request and response models for ``/api/observation``."""

from __future__ import annotations

import datetime
from typing import Any

from pydantic import BaseModel, ConfigDict, Field

from skyportal_py_models.instruments import Instrument, InstrumentField


class Observation(BaseModel):
    """A survey observation (upstream ``ExecutedObservation``/``QueuedObservation``).

    The endpoint returns either kind depending on ``observation_status``,
    so the executed-only fields (``observation_id``, ``airmass``,
    ``seeing``, ``limmag``, ``target_name``, ``processed_fraction``) and
    the queued-only ones (``queue_name``, ``validity_window_start`` and
    ``validity_window_end``) are all optional.
    """

    model_config = ConfigDict(extra="forbid")

    id: int
    created_at: datetime.datetime | None = None
    modified: datetime.datetime | None = None
    instrument_id: int | None = None
    instrument_field_id: int | None = None
    observation_id: int | None = None
    obstime: datetime.datetime | None = None
    filt: str | None = None
    exposure_time: int | None = None
    airmass: float | None = None
    seeing: float | None = None
    limmag: float | None = None
    target_name: str | None = None
    processed_fraction: float | None = None
    queue_name: str | None = None
    validity_window_start: datetime.datetime | None = None
    validity_window_end: datetime.datetime | None = None
    field: InstrumentField | None = None
    instrument: Instrument | None = None


class ObservationsPage(BaseModel):
    """One page of results from an observations query."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    observations: list[Observation] = Field(default_factory=list)
    total_matches: int = Field(alias="totalMatches", default=0)
    probability: float | None = None
    area: float | None = None
    geojson: list[dict[str, Any]] | None = None
    field_ids: list[int] | None = None
    min_observations_per_field: int | None = None


class ObservationPost(BaseModel):
    """Payload for ingesting a set of executed observations."""

    model_config = ConfigDict(extra="forbid", validate_by_name=True)

    telescope_name: str = Field(alias="telescopeName")
    instrument_name: str = Field(alias="instrumentName")
    observation_data: dict[str, list[Any]] = Field(alias="observationData")


class ObservationSimSurveyResponse(BaseModel):
    """Result of starting a SimSurvey efficiency calculation."""

    model_config = ConfigDict(extra="forbid")

    id: int


class ObservationQueues(BaseModel):
    """Queue names retrieved from an instrument's external API."""

    model_config = ConfigDict(extra="forbid")

    queue_names: list[str] = Field(default_factory=list)

File: packages/api-models-ts/src/Observations.ts
/**
 * Request and response models for `/api/observation`.
 *
 * @since 1.0.0
 */

import * as v from "valibot";

import * as Instruments from "./Instruments.ts";
import * as Schemas from "./Schemas.ts";

/**
 * A survey observation (upstream `ExecutedObservation`/`QueuedObservation`).
 *
 * The endpoint returns either kind depending on `observationStatus`, so the
 * executed-only fields (`observation_id`, `airmass`, `seeing`, `limmag`,
 * `target_name`, `processed_fraction`) and the queued-only ones
 * (`queue_name`, `validity_window_start` and `validity_window_end`) are all
 * optional.
 *
 * @since 1.0.0
 * @category Models
 */
export const Observation = Schemas.model(
    v.strictObject({
        id: Schemas.Integer,
        created_at: Schemas.NullishTimestamp,
        modified: Schemas.NullishTimestamp,
        instrument_id: Schemas.NullishInteger,
        instrument_field_id: Schemas.NullishInteger,
        observation_id: Schemas.NullishInteger,
        obstime: Schemas.NullishTimestamp,
        filt: Schemas.NullishString,
        exposure_time: Schemas.NullishInteger,
        airmass: Schemas.NullishNumber,
        seeing: Schemas.NullishNumber,
        limmag: Schemas.NullishNumber,
        target_name: Schemas.NullishString,
        processed_fraction: Schemas.NullishNumber,
        queue_name: Schemas.NullishString,
        validity_window_start: Schemas.NullishTimestamp,
        validity_window_end: Schemas.NullishTimestamp,
        field: Schemas.nullish(Instruments.InstrumentField),
        instrument: Schemas.nullish(Instruments.Instrument),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type Observation = v.InferOutput<typeof Observation>;

/**
 * One page of results from an observations query.
 *
 * @since 1.0.0
 * @category Models
 */
export const ObservationsPage = Schemas.model(
    v.strictObject({
        observations: Schemas.list(Observation),
        totalMatches: v.optional(Schemas.Integer, 0),
        probability: Schemas.NullishNumber,
        area: Schemas.NullishNumber,
        geojson: Schemas.nullish(v.array(Schemas.JsonObject)),
        field_ids: Schemas.nullish(v.array(Schemas.Integer)),
        min_observations_per_field: Schemas.NullishInteger,
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type ObservationsPage = v.InferOutput<typeof ObservationsPage>;

/**
 * Payload for ingesting a set of executed observations.
 *
 * `observationData` maps column names to equal-length lists and must include
 * `observation_id`, `field_id` (or `RA` and `Dec`), `obstime`, `filter`, and
 * `exposure_time`.
 *
 * @since 1.0.0
 * @category Models
 */
export interface ObservationPost {
    readonly telescopeName: string;
    readonly instrumentName: string;
    readonly observationData: Record<string, ReadonlyArray<unknown>>;
}

/**
 * Result of starting a SimSurvey efficiency calculation.
 *
 * @since 1.0.0
 * @category Models
 */
export const ObservationSimSurveyResponse = Schemas.model(v.strictObject({ id: Schemas.Integer }));

/**
 * @since 1.0.0
 * @category Models
 */
export type ObservationSimSurveyResponse = v.InferOutput<typeof ObservationSimSurveyResponse>;

/**
 * Queue names retrieved from an instrument's external API.
 *
 * @since 1.0.0
 * @category Models
 */
export const ObservationQueues = Schemas.model(
    v.strictObject({
        queue_names: Schemas.list(v.string()),
    })
);

/**
 * @since 1.0.0
 * @category Models
 */
export type ObservationQueues = v.InferOutput<typeof ObservationQueues>;
